src/libraries/Microsoft.Extensions.Configuration.Binder/gen/Helpers/Emitter/CoreBindingHelper.cs

```diff
@@ -518,6 +532,30 @@ private void EmitGetBinderOptionsHelper()
                     """);
             }
 
+            private void EmitEnumParseMethod()
+            {
+                string innerExceptionTypeDisplayString = _useFullyQualifiedNames ? "global::System.Exception" : "Exception";
+                string exceptionArg1 = string.Format(ExceptionMessages.FailedBinding, $"{{{Identifier.getPath}()}}", $"{{typeof(T)}}");
+
+                _writer.WriteLine($$"""
+                    public static T ParseEnum<T>(string value, Func<string?> getPath) where T : struct
+                    {
+                        try
+                        {
+                            #if NETFRAMEWORK || NETSTANDARD2_0
+                                return (T)Enum.Parse(typeof(T), value, ignoreCase: true);
+                            #else
+                                return Enum.Parse<T>(value, ignoreCase: true);
```

Comment:
I'd avoid preprocessor directives here. The parser could check/register whether this method exists in the input compilation and then we can emit the right implementation.

Reply:
I'm torn about this. While I don't like that the generated code includes pre-processor directives, at the same time this has the nice property of being consistent in what code is being generated regardless of TFM. This might be useful if you were using the source generator in a multi-targeted project: seeing the same generated source regardless of how you choose to navigate to it in the IDE.

For this particular change I would recommend polyfilling the `Enum.Parse<T>` method for netfx/ns2.0 targets.

Reply:
Makes sense to me.

Reply:
> Adding this kind of branching makes emitter code harder to maintain

Not really: you have a helper in the source generator that emits Enum.Parse, and that one function emits the right variation.  Here we're talking about the difference between emitting `(T)Enum.Parse(typeof(T), ` and `Enum.Parse<T>(`.

> polyfill it

How do you propose polyfilling `Enum.Parse`?

> hide it behind an emitted helper

That would mean always emitting an extra method, even when it's pointless, and that method then contains the call that could have just been at the call site.

We should prioritize emitting the cleanest / best code possible. `Enum.Parse<T>` is both simpler to read and more efficient.

Reply:
> Here we're talking about the difference between emitting (T)Enum.Parse(typeof(T),  and Enum.Parse<T>.

I guess I was trying to make a remark about this problem in the general case. It seems a helper on the emitter that produces the relevant expression would work nicely here. 

I'm mostly concerned about the combinatorial complexity that can arise from a code fragment containing multiple sub-expressions that _could_ avail of new APIs/language features. We should be selective about where and why we do this, using the latest APIs isn't a goal in and of itself. A source generator such as the configuration binder or STJ should ideally just be producing trivial glue code for object graph traversal.

Reply:
Is there any update on what direction (polyfill/check method existance) should I take for my follow up PR ?

Reply:
I believe the consensus is the generate the best code for the target platform - https://github.com/dotnet/runtime/pull/89823#discussion_r1283352188. In general, we'd want to watch for any complexity that may arise with generating different code across TFMs but in this case it's fine.